rfcs/0004-transaction.md

```diff
@@ -0,0 +1,288 @@
+
+1. the `WriteBatch` is committed by `db.write(batch)`, while the `Transaction` is committed by `txn.commit()`
+2. beside the write operations, the `Transaction` should also support the read operations, and if one key is updated in the `Transaction`, the read operation should return the updated value during the transaction.
+
+The `WriteBatch` is considered as a prerequisite for the transaction feature, as it plays as a place to buffer the writes, and it makes MVCC possible with an iterator over it in front of the iterator over Snapshot. So we should implement the `WriteBatch` first before the transaction feature.
+
+We can refer to the WriteBatch RFC or PR for the details later.
+
+## Conflict Checking: Snapshot Isolation
+
+The conflict checking in the SI transaction is simple:
+
+1. Track the sequence number when transaction is started in the `TransactionState` struct.
+2. Track the modified keys in the `TransactionState` struct.
+3. When `commit()` is called, check whether the modified keys are modified by others after the transaction started.
+
+The `TransactionState` struct could be like:
+
+```rust
+struct TransactionState {
+  started_seq: u64,
+  write_keys: HashSet<String>,
+}
+```
+
+How to verify the modified keys are modified by others after the transaction started?
+
+We could simply check the sequence number of the keys in latest db storage. Given a key `"key1"`, let's say the sequence number of the current transaction is "102", while the sequence number of the key `"key1"` is "103" in the storage, then we could regard the key `"key1"` is modified by others after the transaction started, thus violates the conflict check and should abort the commit.
+
+So the problem now is how to get the latest sequence number of a given key in the storage efficiently. It'd not be very fast to read the keys from the disk storage on conflict checking during every commit.
+
+We could reference RocksDB' approach here: RocksDB assumes each transaction is short-lived, and **only checks the conflicts with the changes in recent**. In LSM, the "recent changes" are buffered in MemTable, so we can leverages MemTable to check the conflicts, making conflict checks could be regarded as a pure in-memory operation without any IO overhead.
+
+But the size of MemTable is limited and might be flushed & rotated at any time, it can not buffer ALL the changes since the earliest running transaction. To deal with this limitation, RocksDB's solution is also very straghtforward:
+
+1. Track the earliest sequence number of keys in the MemTable.
+2. When transaction commits, if the MemTable's earliest sequence number is bigger than the transaction's sequence number, then the transaction is considered as `Expired`.
+
+The good part of this approach is that it's very efficient and simple, and it's free of floating garbage. The bad part is that it may abort the transaction when it's not necessary, and prones to encourage the users to set a bigger MemTable size.
```

Comment:
we could consider as a follow-up maintaining a mapping from `key <-> seqnum` for some time even after flushing the memtable to reduce memory pressure of values/attributes while keeping a longer window for valid transactions (not sure if this is something explored in literature or not)

Reply:
yes, the sequence number (seqnum) is also expected to be saved in the SST, every key in both the Memtable and SST will have a sequence number.